### DISCUSSION TITLE
Profiling and Benching

### DISCUSSION THREAD

**Discussion by AliSayyah** [piccolo-orm/piccolo#453] [Ideas]
Hey there!
I want to share some of my ideas for further discussion.

**Profiling**:
We have multiple options:

-  [viztracer](https://github.com/gaogaotiantian/viztracer)
- [pyinstrument](https://github.com/joerick/pyinstrument)
- [py-spy](https://github.com/benfred/py-spy)

They all have an option to generate an interactive webpage from results, and we can serve it with Github pages.
We can have a different repository for this or serve the results in the original docs.
Also, we can have a `profile.sh` script for easy local profiling.

**Benchmarking**:
I found this [action](https://github.com/benchmark-action/github-action-benchmark) for this purpose.
The good thing about it is that we can see the performance regressions and progressions as a chart of commits to the master.
This can also have a script for itself. 

-----------------------
 The common part of profiling and benchmarking is having a python module with all the performance tests and different queries.
And we need to populate a database with a good amount of data for realistic results before testing.

The problem with Github runners is they are shared resources, and the benchmark results will vary a lot. So I don't know how much it makes sense to have workflows for this. But profiling workflow for different platforms could be helpful.
I would appreciate your ideas about this matter to create a task list at the end.

**Comment by dantownsend** [comment]
I think having a local script called `profile.sh` which we can run would be useful.

The GitHub action looks promising. From what I read it gives a high level view (e.g. mean time taken). It could be good for detecting performance regressions.

I think the local script should be more detailed, so we know where the time is being spent, so cprofile, viztracer, pyinstrument etc  might be better for this.

**Comment by AliSayyah** [comment]
Yes, the local profiler would be very useful, but we still have the problem of the #451, where we can't thoughtfully check the performance in other platforms than our local. This could be solved by sharing the profiler results, I think.

**Reply by dantownsend** [reply]
Ideally there would be something like [CodeCov.io](https://app.codecov.io/gh/piccolo-orm/piccolo), where the results could be published for each platform. I don't know of anything like this at the moment though.

**Reply by AliSayyah** [reply]
Yeah, I couldn't find something like this either.

**Reply by dantownsend** [reply]
Yeah, seems like there's a gap in the market! Hopefully we'll find something.

In the mean time, just having the local script would be very useful in case there are non-platform specific performance issues.

**Reply by dantownsend** [reply]
I created a simple proof of concept for a local profiling script:

https://github.com/piccolo-orm/piccolo/pull/454

I'm not saying we have to use this, I've just been learning more about profiling, so it's there for reference in case it's useful.

**Reply by AliSayyah** [reply]
That's great. How did you find `pyinstrument` after using it?

**Reply by dantownsend** [reply]
I've only used it in a very basic way so far. Looking into the docs some more. Seems quite good though.

**Reply by AliSayyah** [reply]
Viztracer has some good features too. Maybe we can compare them both in your sample PR to see which one could give us a better view.

**Reply by dantownsend** [reply]
I'm trying `viztracer` now. Seems very powerful. Has much more information than `pyinstrument`. The UI looks good, but I haven't fully wrapped my head around it yet!

**Reply by AliSayyah** [reply]
I think `viztracer --log_async profiling/profile_select.py` could give you a cleaner view.

**Reply by dantownsend** [reply]
Yeah, that's much better.

**Reply by dantownsend** [reply]
I think `viztracer` is much more useful. I like being able to search in the UI for a particular function, and seeing the code snippet it refers to.

I'll remove `pyinstrument`.

I did find one performance improvement:

https://github.com/piccolo-orm/piccolo/blob/8fd7ec0b95583e56074169b74dd58cab850cb29c/piccolo/query/base.py#L52-L60

It saves around 0.3 ms when fetching 100 rows, which is quite significant.

**Reply by AliSayyah** [reply]
Wow that's great👌

**Reply by dantownsend** [reply]
It's the sort of obscure thing I'd never have realised without a profiling tool!

**Reply by dantownsend** [reply]
Next I want to see if `map` is faster than a list comprehension.

**Reply by AliSayyah** [reply]
I think it depends on the situation. List comprehension could be problematic in case of large data(memory wise)

**Reply by dantownsend** [reply]
Good point

**Reply by dantownsend** [reply]
I tried using `map` - turns out it's lazily evaluated, so it's hard to compare fairly with a list comprehension when profiling. They seem about the same though.

As you say, `map` is more memory efficient because of the lazy evaluation.
